src/hotspot/share/nmt/indexedFreeListAllocator.hpp

```diff
@@ -0,0 +1,149 @@
+#include "memory/allocation.hpp"
+#include "memory/arena.hpp"
+#include "utilities/growableArray.hpp"
+#include "runtime/os.hpp"
+
+// A free list, growth only, allocator for a specific type E.
+// The allocator returns 'pointers' of 4-bytes in size, allowing for
+// memory savings if a pointer-heavy self-referential structure is used.
+// It is "indexed" as a reference is base + index * sizeof(E).
+// It never returns any memory to the system.
+template<typename E, MEMFLAGS flag>
+class IndexedFreeListAllocator {
+public:
+  // Make the index opaque.
+  struct I {
+    int32_t idx;
+    bool operator !=(I other) {
+      return idx != other.idx;
+    }
+    bool operator==(I other) {
+      return idx == other.idx;
+    }
+  };
+  static constexpr const I nil = I{-1};
+  // A free list allocator element is either a link to the next free space
+  // Or an actual element.
+  union BackingElement {
+    I link;
+    E e;
+    BackingElement() {
+      this->link = nil;
+    }
+    BackingElement(I link) {
+      this->link = link;
+    }
+    BackingElement(E& e) {
+      this->e = e;
+    }
+  };
+  GrowableArrayCHeap<BackingElement, flag> backing_storage;
+  I free_start;
+
+  IndexedFreeListAllocator()
+  : backing_storage(8),
+  free_start(I{0}) {}
+
+  template<typename... Args>
+  I allocate(Args... args) {
+    int32_t i = free_start.idx;
+    backing_storage.at_grow(i);
+    BackingElement& be = backing_storage.at(i);
+    if (be.link == nil) {
+      // Must be at end, simply increment
+      free_start.idx += 1;
+    } else {
+      // Follow the link to the next free element
+      free_start = be.link;
+    }
```

Comment:
This feels a bit uncommon and contrary to the normal linked list implementation. If possible, I would opt for a more traditional approach that is easier to understand and does not rely on at_grow (e.g. in case we want adapt this to some other form of backing storage).

e.g.

```
if (_freelist == -1) {
  allocate new slot   
} else {
  reuse slot at freelist
}
```

Reply:
Note that at_grow also could have the problem that an error, resulting in a high index, would cause a large allocation to happen.

I think this class could use an optional max value.